design/one-pager-function-go-templating.md

```diff
@@ -0,0 +1,185 @@
+which expects a map with string keys and base64 encoded string values. This 
+meta resource is not a real Crossplane resource, and it will not be applied to
+the cluster.
+
+```yaml
+apiVersion: meta.gotemplating.fn.crossplane.io/v1alpha1
+kind: CompositeConnectionDetails
+metadata:
+  annotations:
+    gotemplating.fn.crossplane.io/composition-resource-name: connection-details
+  name: connection-details
+{{ if eq $.observed.resources nil }}
+data: {}
+{{ else }}
+data:
+  username: {{ ( index $.observed.resources "sample-access-key-0" ).connectionDetails.username }}
+  password: {{ ( index $.observed.resources "sample-access-key-0" ).connectionDetails.password }}
+  url: {{ "http://www.example.com" | b64enc }}
+{{ end }}
+
+```
+
+## Configuring Resource Status
+
+In addition to that, a new custom annotation, 
+`gotemplating.fn.crossplane.io/ready` will be introduced to set the ready 
+status of the resource, which can be assigned to `Unspecified`, `True` or 
+`False`. This annotation will be removed after the resource is rendered and not
+applied to the cluster.
+
+```yaml
+apiVersion: kubernetes.crossplane.io/v1alpha1
+kind: ProviderConfig
+metadata:
+  name: composed-providerconfig
+  annotations:
+    gotemplating.fn.crossplane.io/ready: True
+spec:
+  credentials:
+    source: Secret
+    secretRef:
+      namespace: crossplane-system
+      name: cluster-config
+      key: kubeconfig
+```
+
+## Helper Functions
+
+A set of helper functions like `getComposedResource` or `getCompositionEnvVar` will be 
+provided to make it easier to access resources or environment config variables 
+referenced in the composition. Without these functions, users can access 
+resources or environment config variables by using the `index` function like 
+below:
+
+```yaml
+{{- range $i := until ( .observed.composite.resource.spec.count | int ) }}
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: User
+metadata:
+  annotations:
+    gotemplating.fn.crossplane.io/composition-resource-name: test-user-{{ $i }}
+  name: test-user-{{ $i }}
+  labels:
+    testing.upbound.io/example-name: test-user-{{ $i }}
+  {{ if eq $.observed.resources nil }}
+    dummy: {{ randomChoice "foo" "bar" "baz" }}
+  {{ else }}
+    dummy: {{ ( index $.observed.resources ( print "test-user-" $i ) ).resource.metadata.labels.dummy }}
+  {{ end }}
+    env: {{ (index $.context "apiextensions.crossplane.io/environment").key1 }}
+spec:
+  forProvider: {}
+{{- end }}
+```
+
+With the helper functions, users can access resources or environment config 
+variables like:
+
+```yaml
+{{- range $i := until ( .observed.composite.resource.spec.count | int ) }}
+---
+apiVersion: iam.aws.upbound.io/v1beta1
+kind: User
+metadata:
+  annotations:
+    gotemplating.fn.crossplane.io/composition-resource-name: test-user-{{ $i }}
+  labels:
+    testing.upbound.io/example-name: test-user-{{ $i }}
+    {{ $composed := getComposedResource $.observed ( print "test-user-" $i ) }}
```

Comment:
just to clarify, the `getComposedResource ` can access any of the composed resources involved in Composition?

Reply:
`getComposedResource` will take two parameters. The first one is the `observed` map which contains **all of the** composed resources. The second parameter is the key which is equal to the value of `gotemplating.fn.crossplane.io/composition-resource-name` annotation.